typing ir.py - Disallow untyped defs for ir.py

aorenste opened the pull request:
- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()
- `ruff check torch/_inductor/ir.py --select ANN --fix --unsafe-fixes` and then fixed up things that looked incorrectly applied.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131846
* __->__ #139238



cc @ezyang @malfet @xuzhao9 @gramster @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

aorenste committed:
typing ir.py - Disallow untyped defs for ir.py

[ghstack-poisoned]

aorenste committed:
Update on "typing ir.py - Disallow untyped defs for ir.py"

- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/ir.py:
```diff
@@ -831,20 +833,20 @@ def store_reduction(self, output_name, indexer, vars, reduction_vars):
         return ops.store_reduction(output_name, indexer(vars), value)
 
-    def index_length(self):
+    def index_length(self):  # type: ignore[no-untyped-def]
```
```suggestion
    def index_length(self) -> int:
```

Skylion007 commented on torch/_inductor/ir.py:
```diff
@@ -1,4 +1,3 @@
-# mypy: allow-untyped-defs
```
Use ruff ANN rules to infer trivial return type annotations here

aorenste committed:
Update on "typing ir.py - Disallow untyped defs for ir.py"

- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()
- `ruff check torch/_inductor/ir.py --select ANN --fix --unsafe-fixes` and then fixed up things that looked incorrectly applied.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ezyang reviewed (approved):
The "add per line suppresses" is not a very common strategy in our codebase but I guess it works

aorenste commented:
> The "add per line suppresses" is not a very common strategy in our codebase but I guess it works

I have a stack of PRs which I'm slowly landing in the background which removes a lot of these per-line suppressions.  My general philosophy is to set the flags to prevent new missing types from being accidentally landed and then backfill the existing missing types.

ezyang commented:
The main problem is that it is difficult to distinguish intentional suppressions from

Skylion007 commented:
> The main problem is that it is difficult to distinguish intentional suppressions from automatically added ones. This is less of a problem if you immediately fix them

We can add some other pragma that these are autogenerated, it only has to have type:ignore in the regex if I recall.

Skylion007 commented:
@pytorchbot merge

aorenste commented:
For my info: what's an example or case where we add an automatically added suppression? Are these added and expected to be permanent additions or is it more like "some PR broke lint and we're fixing it forward but you should fix this"?

aorenste committed:
Update on "typing ir.py - Disallow untyped defs for ir.py"

- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()
- `ruff check torch/_inductor/ir.py --select ANN --fix --unsafe-fixes` and then fixed up things that looked incorrectly applied.




cc ezyang malfet xuzhao9 gramster voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "typing ir.py - Disallow untyped defs for ir.py"

- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()
- `ruff check torch/_inductor/ir.py --select ANN --fix --unsafe-fixes` and then fixed up things that looked incorrectly applied.




cc ezyang malfet xuzhao9 gramster voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge

aorenste committed:
Update on "typing ir.py - Disallow untyped defs for ir.py"

- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()
- `ruff check torch/_inductor/ir.py --select ANN --fix --unsafe-fixes` and then fixed up things that looked incorrectly applied.




cc ezyang malfet xuzhao9 gramster voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
None on "typing ir.py - Disallow untyped defs for ir.py"

- Remove "mypy: allow-untyped-defs" and mark functions individually with "no-untyped-def"
- Mark some trivial functions with the proper return types (`None` and `torch.dtype`)
- Fixed a type bug in the signature of supported_dtype_of_cpp_wrapper()
- `ruff check torch/_inductor/ir.py --select ANN --fix --unsafe-fixes` and then fixed up things that looked incorrectly applied.




cc ezyang malfet xuzhao9 gramster voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

malfet commented on torch/_inductor/ir.py:
```diff
@@ -6672,11 +6675,11 @@ def should_realize_on_reuse(self, users):
         return False
 
-    def mark_reuse(self, users):
+    def mark_reuse(self, users) -> None:  # type: ignore[no-untyped-def]
```
```suggestion
    def mark_reuse(self, users) -> None:
```

malfet commented on torch/_inductor/ir.py:
```diff
@@ -6648,13 +6651,13 @@ def realize_hint(self):
         )
 
-    def should_realize_on_reuse(self, users):
+    def should_realize_on_reuse(self, users):  # type: ignore[no-untyped-def]
```
```suggestion
    def should_realize_on_reuse(self, users) -> bool:
```

malfet commented on torch/_inductor/ir.py:
```diff
@@ -6648,13 +6651,13 @@ def realize_hint(self):
             self.realize()
 
-    def has_exceeded_max_reads(self):
+    def has_exceeded_max_reads(self):  # type: ignore[no-untyped-def]
```
```suggestion
    def has_exceeded_max_reads(self) -> bool:
```

malfet commented on torch/_inductor/ir.py:
```diff
@@ -6771,7 +6774,7 @@ def handle_sym_expr(stride):
         return outputs
 
-    def codegen(self, wrapper):
+    def codegen(self, wrapper) -> None:  # type: ignore[no-untyped-def]
```
```suggestion
    def codegen(self, wrapper) -> None:
```

malfet reviewed (changes requested):
You have an anti pattern of
```
def foo() -> None:  # type: ignore[no-untyped-def]
```
Please fix it (suggested a few changes)
Also, if you want to enforce it, you'll need to modify lint runner.yaml, don't you?

aorenste commented:
> You have an anti pattern of
> 
> ```
> def foo() -> None:  # type: ignore[no-untyped-def]
> ```
> 
> Please fix it (suggested a few changes) Also, if you want to enforce it, you'll need to modify lint runner.yaml, don't you?

I think you're confused.  The purpose of this is to get us to a better state:
1. Disallow untyped defs by default and mark broken files individually (this was done in #127838)
2. When fixing a given file start by removing the file-level ignore and replace with function-level ignores (this PR)
3. Fix individual functions removing their function-level ignores. This is done in the rest of this stack (I did this during the last BE week but haven't had a lot of time to land) - this is the stack ending with #131852

The problem is that any one change can cause a ripple effect affecting MANY files so it's a lot easier to land smaller changes with temporary ignores than trying to land an single large change.
